pdns/dns_random.cc

```diff
@@ -22,73 +22,263 @@
-  memcpy(g_counter, &now.tv_usec, sizeof(now.tv_usec));
-  memcpy(g_counter+sizeof(now.tv_usec), &now.tv_sec, sizeof(now.tv_sec));
-  g_in = getpid() | (getppid()<<16);
+# if defined(HAVE_GETRANDOM)
+  if (chosen_rng == RNG_GETRANDOM) {
+    char buf[1];
+    // some systems define getrandom but it does not really work, e.g. because it's
+    // not present in kernel.
+    if (getrandom(buf, sizeof(buf), 0) == -1) {
+       L<<Logger::Warning<<"getrandom() failed: "<<strerror(errno)<<", falling back to " + rdev<<std::endl;
+       chosen_rng = RNG_URANDOM;
+    }
+  }
+# endif
+
+#if defined(HAVE_RAND_BYTES)
+  if (chosen_rng == RNG_OPENSSL) {
+    int ret;
+    unsigned char buf[1];
+    if ((ret = RAND_bytes(buf, sizeof(buf))) == -1)
+      throw PDNSException("RAND_bytes not supported by current SSL engine");
+    if (ret == 0)
+      throw PDNSException("Openssl RNG was not seeded");
+  }
+# endif
+#endif /* USE_URANDOM_ONLY */
+  if (chosen_rng == RNG_URANDOM) {
+    urandom_fd = open(rdev.c_str(), O_RDONLY);
+    if (urandom_fd == -1)
+      throw PDNSException("Cannot open " + rdev + ": " + std::string(strerror(errno)));
+  }
 
-  g_initialized = true;
-  srandom(dns_random(numeric_limits<uint32_t>::max()));
+  if (chosen_rng == RNG_KISS) {
+    unsigned int seed;
+    urandom_fd = open(rdev.c_str(), O_RDONLY);
+    if (urandom_fd == -1)
+      throw PDNSException("Cannot open " + rdev + ": " + std::string(strerror(errno)));
+    if (read(urandom_fd, &seed, sizeof(seed)) < 0) {
+      (void)close(urandom_fd);
+      throw PDNSException("Cannot read random device");
+    }
+    kiss_init(seed);
+    (void)close(urandom_fd);
+  }
 }
 
-unsigned int dns_random(unsigned int n)
-{
-  if(!g_initialized)
-    abort();
-  uint32_t out;
-#if OPENSSL_VERSION_NUMBER > 0x1010000fL && !defined LIBRESSL_VERSION_NUMBER
-  CRYPTO_ctr128_encrypt((const unsigned char*)&g_in, (unsigned char*) &out, sizeof(g_in), &aes_key, g_counter, g_stream, &g_offset, (block128_f) AES_encrypt);
-#else
-  AES_ctr128_encrypt((const unsigned char*)&g_in, (unsigned char*) &out, sizeof(g_in), &aes_key, g_counter, g_stream, &g_offset);
-#endif
-  return out % n;
+void dns_random_init(const string& data) {
+  unsigned int seed;
+  dns_random_setup();
+  (void)dns_random(1);
+  // init should occur already in dns_random_setup
+  // this interface is only for KISS
+  if (chosen_rng != RNG_KISS)
+    return;
+  if (data.size() != 16)
+    throw PDNSException("invalid seed");
+  seed = (data[0] + (data[1]<<8) + (data[2]<<16) + (data[3]<<24)) ^
+         (data[4] + (data[5]<<8) + (data[6]<<16) + (data[7]<<24)) ^
+         (data[8] + (data[9]<<8) + (data[10]<<16) + (data[11]<<24)) ^
+         (data[12] + (data[13]<<8) + (data[14]<<16) + (data[15]<<24));
+  kiss_init(seed);
 }
 
-#if 0
-int main()
-{
-  dns_random_init("0123456789abcdef");
+/* Parts of this code come from arc4random_uniform */
```

Comment:
You might want to include the comments about "modulo bias" too, since that explain why you bother computing a lower threshold and looping until the number is ok.

Reply:
added comment.